include/soci/rowset.h

```diff
@@ -136,7 +141,7 @@ class rowset_impl
     iterator begin() const
     {
         // No ownership transfer occurs here
-        return iterator(*st_, *define_);
+        return iterator(st_.get(), define_.get());
```

Comment:
I don't understand how does this not result in a crash... `rowset_iterator` ctor calls `++` which uses the pointer, which is null here for an invalid rowset.

Am I missing something here or would it actually crash if you added a test exercising the new code?

Reply:
I have been testing this and it does work. However, I also wonder why the `++(*this)` doesn't make it explode... I can do a gdb session on it and see where it takes me.

Talking about the design itself, I agree that changing the `rowset_iterator` ctor is not ideal, however, I was taken away by the beauty of the minimal change to `begin()`. I did another version in which I didn't change the `rowset_iterator` instead I had:

`iterator begin() const`
`{`
`    // No ownership transfer occurs here`
`    return (st_.get() == nullptr)`
`        ? iterator()`
`        : iterator(*st_, *define_);`
`}`

Reply:
Well, beauty is in the eye of the beholder and personally I find a ctor taking optional and maybe null pointers pretty ugly. And

```cpp
iterator begin() const
{
    // Empty rowset doesn't have any valid begin iterator.
    return st_ ? iterator(*st_, *define_) : iterator();
}
```

seems perfectly fine to me.

Reply:
It's coming up - I just need to test it - while I'm at it, should I add a clear for #198 and close a 10 year old issue? ;)